examples/virus_antibody/agents.py

```diff
@@ -0,0 +1,259 @@
+    def __init__(
+        self,
+        model,
+        space,
+        sight_range,
+        duplication_rate,
+        ko_timeout,
+        memory_capacity,
+        initial_position=(0, 0),
+        direction=(1, 1),
+    ):
+        super().__init__(model=model, space=space)
+
+        # Movement & state
+        self.position = initial_position
+        self.speed = 1.5
+        self.direction = np.array(direction, dtype=float)
+
+        # Characteristics
+        self.sight_range = sight_range
+        self.health = 2
+        self.duplication_rate = duplication_rate
+
+        # Memory
+        self.st_memory: deque = deque()
+        self.lt_memory: list = []
+        self.memory_capacity = memory_capacity
+
+        # Target & KO state
+        self.target = None  # will hold a weakref.ref or None
+        self.ko_timeout = ko_timeout
+        self.ko_steps_left = 0
+
+    def step(self):
+        # Acquire a virus target if we don't already have one
+        if self.target is None:
+            closest = self.find_closest_virus()
+            if closest:
+                self.target = weakref.ref(closest)
+
+        # Communicate and maybe duplicate
+        self.communicate()
+        if self.random.random() < self.duplication_rate:
+            self.duplicate()
+
+        # Then move
+        self.move()
+
+    def find_closest_virus(self):
+        agents, _ = self.space.get_agents_in_radius(self.position, self.sight_range)
+        viruses = [a for a in agents if isinstance(a, VirusAgent)]
+        return viruses[0] if viruses else None
+
+    def communicate(self) -> bool:
+        agents, _ = self.space.get_agents_in_radius(self.position, self.sight_range)
+        peers = [
+            a
+            for a in agents
+            if isinstance(a, AntibodyAgent) and a.unique_id != self.unique_id
+        ]
+        if not peers:
+            return False
+
+        for other in peers:
+            to_share = [
+                dna for dna in self.st_memory if dna and dna not in other.lt_memory
+            ]
+            if to_share:
+                other.st_memory.extend(to_share)
+                other.lt_memory.extend(to_share)
+                while len(other.st_memory) > self.memory_capacity:
+                    other.st_memory.popleft()
+        return True
+
+    def duplicate(self):
+        clone = AntibodyAgent(
+            self.model,
+            self.space,
+            sight_range=self.sight_range,
+            duplication_rate=self.duplication_rate,
+            ko_timeout=self.ko_timeout,
+            memory_capacity=self.memory_capacity,
+            initial_position=self.position,
+            direction=self.direction,
+        )
+        # Copy over memory
+        clone.st_memory = deque(item for item in self.st_memory if item)
+        clone.lt_memory = [item for item in self.lt_memory if item]
+        clone.target = None
+        clone.ko_steps_left = 0
+
+        self.model.antibodies_set.add(clone)
```

Comment:
this should not be needed. `model.agents_by_type[AntibodyAgent]` will handle this already for you.